### PR TITLE
SPLICE-1591: purge deleted rows during major compaction

### PR DESCRIPTION
**Pull request by jyuanca**
backport to 2.5

### CODE REVIEW

**Comment by dgomezferro** [comment]
@jyuanca , I upgraded a 2.5 standalone version and it came up correctly, I see the added column log:

````
11:54:34,714 (pool-15-thread-1) INFO  [c.s.d.i.s.c.SpliceDataDictionary] - SYS.SYSTABLES upgraded: added a new column PURGE_DELETED_ROWS.
`````
but I got the following error when trying to select from sys.systables:

````
splice> select * from sys.systables;
ERROR XJ001: DERBY SQL error: SQLCODE: -1, SQLSTATE: XJ001, SQLERRMC: java.lang.IllegalArgumentExceptionBit position -1 is outside the legal rangeXJ001.U
````

splice-derby.log:

````
848000569140316{3}), Failed Statement is: select * from sys.systables
java.lang.IllegalArgumentException: Bit position -1 is outside the legal range
        at com.splicemachine.db.iapi.services.io.FormatableBitSet.checkPosition(FormatableBitSet.java:88)
        at com.splicemachine.db.iapi.services.io.FormatableBitSet.set(FormatableBitSet.java:543)
        at com.splicemachine.db.impl.sql.compile.ResultColumnList.getReferencedFormatableBitSet(ResultColumnList.java:3239)
        at com.splicemachine.db.impl.sql.compile.FromBaseTable.changeAccessPath(FromBaseTable.java:1671)
        at com.splicemachine.db.impl.sql.compile.ProjectRestrictNode.modifyAccessPath(ProjectRestrictNode.java:634)
        at com.splicemachine.db.impl.sql.compile.OptimizerImpl.modifyAccessPaths(OptimizerImpl.java:1204)
        at com.splicemachine.db.impl.sql.compile.SelectNode.modifyAccessPaths(SelectNode.java:1618)
        at com.splicemachine.db.impl.sql.compile.DMLStatementNode.optimizeStatement(DMLStatementNode.java:152)
        at com.splicemachine.db.impl.sql.compile.CursorNode.optimizeStatement(CursorNode.java:382)
        at com.splicemachine.db.impl.sql.GenericStatement.bindAndOptimize(GenericStatement.java:584)
        at com.splicemachine.db.impl.sql.GenericStatement.fourPhasePrepare(GenericStatement.java:477)
        at com.splicemachine.db.impl.sql.GenericStatement.prepMinion(GenericStatement.java:423)
        at com.splicemachine.db.impl.sql.GenericStatement.prepare(GenericStatement.java:108)
        at com.splicemachine.db.impl.sql.conn.GenericLanguageConnectionContext.prepareInternalStatement(GenericLanguageConnectionContext.java:1087)
        at com.splicemachine.db.impl.jdbc.EmbedPreparedStatement.<init>(EmbedPreparedStatement.java:120)
        at com.splicemachine.db.impl.jdbc.EmbedPreparedStatement20.<init>(EmbedPreparedStatement20.java:73)
        at com.splicemachine.db.impl.jdbc.EmbedPreparedStatement30.<init>(EmbedPreparedStatement30.java:67)
        at com.splicemachine.db.impl.jdbc.EmbedPreparedStatement40.<init>(EmbedPreparedStatement40.java:48)
        at com.splicemachine.db.jdbc.Driver40.newEmbedPreparedStatement(Driver40.java:107)
        at com.splicemachine.db.impl.jdbc.EmbedConnection.prepareStatement(EmbedConnection.java:1210)
        at com.splicemachine.db.impl.jdbc.EmbedConnection.prepareStatement(EmbedConnection.java:1091)
        at com.splicemachine.db.impl.drda.DRDAStatement.prepare(DRDAStatement.java:686)
        at com.splicemachine.db.impl.drda.DRDAStatement.explicitPrepare(DRDAStatement.java:647)
        at com.splicemachine.db.impl.drda.DRDAConnThread.parsePRPSQLSTT(DRDAConnThread.java:3927)
        at com.splicemachine.db.impl.drda.DRDAConnThread.processCommands(DRDAConnThread.java:870)
        at com.splicemachine.db.impl.drda.DRDAConnThread.run(DRDAConnThread.java:335)
Cleanup action completed

````

**Comment by jyuanca** [comment]
@dgomezferro Storage position was not correctly set. The problem has been fixed. Please take another look. Thanks!

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/SpliceDataDictionary.java [hunk 1]
```diff
@@ -522,8 +512,51 @@ private void upgradeIfNecessary(TransactionController tc) throws StandardExcepti
             tc.setProperty(SPLICE_DATA_DICTIONARY_VERSION,spliceSoftwareVersion,true);
             tc.commit();
         }
+        // TODO - include the following in an upgrade script
+        if (spliceSoftwareVersion.toLong() >= 2005000) {
```
Comment (snowspeeder) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/SpliceDataDictionary.java]: can you please define constant for the version number here

**Comment by dgomezferro** [comment]
@jyuanca  it worked now, thanks.

There's a small problem where if the specified table doesn't exist for the purge deleted rows procedure it returns a Null pointer exception, rather than saying the table doesn't exist.

Also, the procedure doesn't invalidate the cached table descriptors, so if the table was previously major compacted, and then purge deleted rows is set to true, the next major compaction might not delete the rows if the table descriptor is still cached.

**Comment by wtang16** [comment]
@jyuanca Could you address Daniel's comments?

**Pull request merged**